controller/src/main/java/org/jboss/as/controller/ValidateModelStepHandler.java

```diff
@@ -147,9 +183,81 @@ public void execute(OperationContext context, ModelNode operation) throws Operat
 
     }
 
+    private void handleObjectAttributes(ModelNode model, AttributeDefinition attr, String absoluteParentName) throws OperationFailedException {
+        if (attr instanceof ObjectTypeAttributeDefinition) {
+            validateNestedAttributes(model.get(attr.getName()), (ObjectTypeAttributeDefinition) attr, absoluteParentName);
+        } else if (attr instanceof ObjectListAttributeDefinition) {
+            ObjectTypeAttributeDefinition valueType = ((ObjectListAttributeDefinition) attr).getValueType();
+            ModelNode list = model.get(attr.getName());
+            for (int i = 0; i < list.asInt(); i++) {
+                validateNestedAttributes(list.get(i), valueType, absoluteParentName + "[" + i + "]");
+            }
+        } else if (attr instanceof ObjectMapAttributeDefinition) {
+            ObjectTypeAttributeDefinition valueType = ((ObjectMapAttributeDefinition) attr).getValueType();
+            ModelNode map = model.get(attr.getName());
+            for (String key : map.keys()) {
+                validateNestedAttributes(map.get(key), valueType, absoluteParentName + "." + key);
+            }
+        }
+    }
+
+    private void validateNestedAttributes(ModelNode subModel, ObjectTypeAttributeDefinition attr, String absoluteParentName) throws OperationFailedException {
+        if (!subModel.isDefined()) {
+            return;
+        }
+        AttributeDefinition[] subAttrs = attr.getValueTypes();
+        Map<String, AttributeDefinition> visited = new HashMap<>();
+        for (AttributeDefinition subAttr : subAttrs) {
+            String subAttributeName = subAttr.getName();
+            visited.put(subAttributeName, subAttr);
+            String absoluteName = absoluteParentName + "." + subAttributeName;
+            if (!subModel.hasDefined(subAttributeName) && isRequired(subAttr, subModel)) {
+                throw new OperationFailedException(ControllerLogger.MGMT_OP_LOGGER.required(subAttributeName));
+            }
+            if (!subModel.hasDefined(subAttributeName)) {
+                continue;
+            }
+            String[] requires = subAttr.getRequires();
+            if (requires != null) {
+                for (final String required : requires) {
```

Comment:
what happens if requires is defined with dot notation, aka having more than one level of nesting?